### PR TITLE
bpo-35224: PEP 572 Implementation

### PR DESCRIPTION
**Pull request by emilyemorehouse**
This is the implementation of PEP 572.  It has been few a couple of rounds of refactoring per my and Guido's review, but is now ready for a more thorough review from others.

The only TODO items before merging are:

- [x] News entry
- [x] Thorough reviews 
---
### Additional Notes
The new test cases of note can be explicitly run using: 

    ./python.exe -m test -v test_parser -v test_named_expressions
    ./python.exe -m test -v test_parser -v test_parser
    ./python.exe -m test -v test_parser -v test_tokenize

#### Please note the following caveats and future tasks. These items are NOT blocking reviews or merging, but are here as FYIs if they come up in review:
* ~Must decide upon cleaner implementation of disallowing expressions such as `((a) := 10)`, or to continuing allowing this. The current, and know-to-be-imperfect, implementation was introduced [here](https://github.com/python/cpython/commit/f1531400ca7d7a2d148830c8ac703f041740896d#diff-4d35cf8992b795c5e97e9c8b6167cb34R1718).~ This change has been reverted and deferred
* There are a few minor differences between how the implementation functions versus the spec in the PEP. See below for more information.
* Documentation updates will come in a separate PR

##### Differences between the PEP spec and implementation

From the "Scope of the target" section of the PEP, there are two cases that should raise a TargetScopeError: when an assignment expression is used in a comprehension inside a class body or for special cases in comprehensions.

Invalid examples for the latter include:

    [i := i+1 for i in range(5)]
    [[(j := j) for i in range(5)] for j in range(5)]
    [i := 0 for i, j in stuff]
    [i+1 for i in i := stuff]

However, the following work in the implementation,though the PEP states they should be invalid:

    >>> [i := i+1 for i in range(5)]
    [1, 2, 3, 4, 5]
    >>> i
    5

    >>> [i := 0 for i, j in [(1, 2)]]
    [0]

The following does not work in the implementation (as desired), but does not throw a TargetScopeError as defined in the PEP:

    >>> [i+1 for i in i := range(5)]
    File "<stdin>", line 1
        [i+1 for i in i := range(5)]
                        ^
    SyntaxError: invalid syntax

https://bugs.python.org/issue35224

### CODE REVIEW

File: Python/ast.c
Comment (emilyemorehouse) [inline Python/ast.c]: This section through L1760 is a known issue and is marked for refactoring (see PR description.)

**Comment by tirkarthi** [comment]
> Must decide upon cleaner implementation of disallowing expressions such as ((a) := 10), or to continuing allowing this. The current, and know-to-be-imperfect, implementation was introduced here.

There was some discussion around disallowing `(a) = 10` but then was considered as valid in the end. I couldn't find the relevant part of PEP 572 about this scenario but just wanted to add this here for reference : https://bugs.python.org/issue35207 . Maybe this can be added to the PEP once it's decided as a valid or invalid syntax.

**Comment by vstinner** [comment]
@emilyemorehouse: Can you please add a NEWS entry using blurb?

Travis CI failed on "FAIL: test_exceptions (test.test_pickle.CompatPickleTests) [TargetScopeError]", I didn't check why yet.

**Comment by vstinner** [comment]
> Please note the following caveats and TODO items: (...)

Ah, it seems like you don't want to merge this PR in its current state, so I took the liberty of adding [WIP] to the PR title and the DO-NOT-MERGE label.

**Comment by gvanrossum** [comment]
I hope the [WIP] label, DO-NOT-MERGE label and presence of caveats and
TODOs don't prevent qualified core devs from reviewing this PR. In order to
be able to experiment with the feature at all we need some baseline, and
this PR is meant to be that. In some cases we may find that the behavior of
the code is just fine and it's better to tweak the PEP. In other cases we
may want to postpone solving something and plan to follow up with another
PR that tweaks some edge case (e.g. to give better diagnostics). In yet
other cases we need to fix this PR before it can be merged (e.g. if there
are refcount bugs or other crashables).

**Comment by emilyemorehouse** [comment]
@gvanrossum Agreed. I'll update the PR description to reflect this as well, but I'm leaning towards (in no particular order):

* Fix CI failures
* Revert current `((a) := 10)` restriction, to be decided later
* Thorough reviews

Then merge. Anything else can be handled in subsequent PRs.

**Comment by emilyemorehouse** [comment]
(But please, either leave the `do-not-merge` label, or be aware that I'd like to merge this PR myself)

**Comment by gvanrossum** [comment]
Yes, I'd focus on news entry, reverting current ((a) := 10) restriction, and CI failures -- and thorough reviews. (Serhiy?)

**Comment by gvanrossum** [comment]
@tirkarthi 
>  I couldn't find the relevant part of PEP 572 about this scenario

This phrase in the PEP is normative:

>  This is of the form `NAME := expr` where `expr` is any valid Python expression other than an unparenthesized tuple, and `NAME` is an identifier.

Note that unlike regular assignment, PEP 572 does not allow any form of tuple unpacking. This is intentional: we couldn't find any examples where it would actually be useful.

**Comment by gvanrossum** [comment]
I removed the DO_NOT_MERGE label -- if this passes review I would be happy to see it merged (but please let Emily merge it herself). Can we get a review from @benjaminp or @vstinner ? Also @serhiy-storchaka if you can take a look that would be appreciated.

File: .gitattributes
Comment (serhiy-storchaka) [inline .gitattributes]: These files are not generated.

File: Grammar/Grammar
Comment (serhiy-storchaka) [inline Grammar/Grammar:112]: Does not `[x := 1, 2]` look *too* ambiguous? Why this syntax is allowed? `{x := 1, 2}`, `{x := 1 : 2}` and `{1 : x := 2}` are not allowed.

File: Grammar/Grammar
Comment (serhiy-storchaka) [inline Grammar/Grammar:138]: This looks like a bug magnet: `f(x=1)` vs `f(x:=1)`. Is this desired?

File: Lib/tokenize.py
Comment (serhiy-storchaka) [inline Lib/tokenize.py]: This adds the operator `:`.

File: Parser/Python.asdl
Comment (serhiy-storchaka) [inline Parser/Python.asdl:91]: Why `NamedStore` is needed? Can not `Store` be used instead?

File: Python/ast.c
Comment (serhiy-storchaka) [inline Python/ast.c:1745]: Seems `(x) := 1` is accepted. Is it desirable? If it is not, see `ast_for_call` which makes `f(x=1)` allowed, but rejects `f((x)=1)`.

File: Python/ast.c
Comment (serhiy-storchaka) [inline Python/ast.c]: PEP-7: Missed space before `(`. Also add braces around `if`s body, even if it is just `return`.

File: Python/ast.c
Comment (serhiy-storchaka) [inline Python/ast.c:2929]: I prefer to avoid code duplication. Maybe merge this case with the case for `NCH(ch) == 1`?
```c
if (NCH(ch) == 1 || (NCH(ch) == 3 && TYPE(CHILD(ch, 1)) == COLONEQUAL)) {
    if (nkeywords) {
        ...
    }
}
```

File: Python/compile.c
Comment (serhiy-storchaka) [inline Python/compile.c]: Is not `x.a := expr` forbidden?

File: Python/compile.c
Comment (serhiy-storchaka) [inline Python/compile.c]: Is not `x[i] := expr` forbidden?

File: Python/compile.c
Comment (serhiy-storchaka) [inline Python/compile.c]: How is it possible?

**Review by serhiy-storchaka — COMMENTED** [review]
This is a preliminary review.

**Comment by vstinner** [comment]
What is the status of this PR? Is it ready to be merged? I recall that Serhiy changed how generated files are created and adjusted .gitattributes. Was it the last blocker issue?

**Comment by tirkarthi** [comment]
@vstinner Serhiy has left some comments and I doubt they will be able to review this given health issues https://mail.python.org/pipermail/python-dev/2019-January/156080.html

**Comment by vstinner** [comment]
> @vstinner Serhiy has left some comments and I doubt they will be able to review this given health issues https://mail.python.org/pipermail/python-dev/2019-January/156080.html

I know. But @emilyemorehouse marked most comments as resolved in the meanwhile.

**Comment by emilyemorehouse** [comment]
As an update, I'm actively working on updating this PR to work with the new changes from [bpo-30455](https://bugs.python.org/issue30455) 🙂 After that, it should be ready to merge

**Comment by vstinner** [comment]
> As an update, I'm actively working on updating this PR to work with the new changes from bpo-30455 slightly_smiling_face After that, it should be ready to merge

Cool. It's nice to see enhancements on these generated files.

**Comment by emilyemorehouse** [comment]
Updated to work with the changes from [bpo-30455](https://bugs.python.org/issue30455), now to update per Ivan's changes in [bpo-33416](https://bugs.python.org/issue33416) 😅

File: Misc/NEWS.d/next/Core and Builtins/2018-11-13-14-26-54.bpo-35224.F0B6UQ.rst
Comment (vstinner) [inline Misc/NEWS.d/next/Core and Builtins/2018-11-13-14-26-54.bpo-35224.F0B6UQ.rst]: Maybe remove "Documentation and edge cases will be issued in separate PRs."

**Comment by vstinner** [comment]
I'm sorry, I pushed changes to rework _PyExc_Init() and now you have to handle conflicts. Would you mind to update your PR again?

Moreover, it seems like you have to regenerate Include/Python-ast.h or rebase your PR (or merge master into your PR).

**Review by vstinner — APPROVED** [review]
LGTM. It seems that I don't see any obvious issue, but I'm unable to review properly the compiler/grammar changes. IMHO the most important part of the PR is the addition of the large test case for PEP 572. Even if we rewrite the implementation tomorrow, as soon as test_named_expressions.py pass, I'm fine with any implementation :-)

I really want to see this change merged into master as soon as possible, to get enough time to play with it and discover any possible corner cases which might need to update the PEP and/or the implementation. I'm fine with pushing documentation changes in a second step, we commonly do that for new features (ex: dataclasses only got its doc a few days just before 3.7 beta1 release :-)).

It seems like other changes to prepare this implementation have already been merged, especially changes on how generated parser/grammar files are generated.

I would prefer to merge this PR as soon as soon also to avoid @emilyemorehouse to have to handle conflicts frequently which is painful for everybody.

@emilyemorehouse: I would suggest you to merge your PR as soon as you fixed conflicts and the CI pass. Be careful with the commit message. I suggest to use a short commit message like:

"Implement PEP 572 (assignment expressions)."

**Review by gvanrossum — APPROVED** [review]
I'm with Victor. Let's land this!

**Comment by emilyemorehouse** [comment]
Thanks, @vstinner, but I was sitting on some changes that just needed testing 🙂  I can include the compiler warning updates if you want. I'd like to merge this one myself.

**Comment by vstinner** [comment]
> I'd like to merge this one myself.

Sure. Go ahead!

> I can include the compiler warning updates if you want.

I'm not sure about these changes, but I would prefer to not introduce new compiler warnings if possible.

**Comment by vstinner** [comment]
Oh, when I proposed to push directly into your branch, I forgot to give the context: Python 3.8a1 is coming soon, so it's time to merge this PR ;-)
https://discuss.python.org/t/python-3-8-0a1-postponed-for-a-week/766/2

It would be great to have it in the first 3.8 (alpha) release ;-)

**Comment by vstinner** [comment]
I'm not sure if all compilers emit the same warnings, so here are the GCC 8.1. warnings:

```
...
gcc -pthread -c -Wno-unused-result -Wsign-compare -g -Og -Wall -O0   -std=c99 -Wextra -Wno-unused-result -Wno-unused-parameter -Wno-missing-field-initializers -Werror=implicit-function-declaration  -I./Include/internal  -I. -I./Include    -DPy_BUILD_CORE -o Python/ceval.o Python/ceval.c
Python/ast.c: In function 'ast_for_expr':
Python/ast.c:2662:16: warning: this statement may fall through [-Wimplicit-fallthrough=]
             if (NCH(n) == 3)
                ^
Python/ast.c:2664:9: note: here
         case test:
         ^~~~
gcc -pthread -c -Wno-unused-result -Wsign-compare -g -Og -Wall -O0   -std=c99 -Wextra -Wno-unused-result -Wno-unused-parameter -Wno-missing-field-initializers -Werror=implicit-function-declaration  -I./Include/internal  -I. -I./Include    -DPy_BUILD_CORE -o Python/codecs.o Python/codecs.c
gcc -pthread -c -Wno-unused-result -Wsign-compare -g -Og -Wall -O0   -std=c99 -Wextra -Wno-unused-result -Wno-unused-parameter -Wno-missing-field-initializers -Werror=implicit-function-declaration  -I./Include/internal  -I. -I./Include    -DPy_BUILD_CORE -o Python/compile.o Python/compile.c
gcc -pthread -c -Wno-unused-result -Wsign-compare -g -Og -Wall -O0   -std=c99 -Wextra -Wno-unused-result -Wno-unused-parameter -Wno-missing-field-initializers -Werror=implicit-function-declaration  -I./Include/internal  -I. -I./Include    -DPy_BUILD_CORE -o Python/coreconfig.o Python/coreconfig.c
Python/compile.c: In function 'compiler_handle_subscr':
Python/compile.c:5017:5: warning: enumeration value 'NamedStore' not handled in switch [-Wswitch]
     switch (ctx) {
     ^~~~~~
gcc -pthread -c -Wno-unused-result -Wsign-compare -g -Og -Wall -O0   -std=c99 -Wextra -Wno-unused-result -Wno-unused-parameter -Wno-missing-field-initializers -Werror=implicit-function-declaration  -I./Include/internal  -I. -I./Include    -DPy_BUILD_CORE -o Python/dynamic_annotations.o Python/dynamic_annotations.c
...
```

**Comment by vstinner** [comment]
> Ah, I only had the switch warning. I added your suggestion for the fallthrough, hopefully that fixes it since I cannot test it.

I confirm that "make" no longer emit warning with your merged change. Good.

**Pull request merged**
